applications/ShallowWaterApplication/python_scripts/shallow_water_base_solver.py

```diff
@@ -1,104 +1,124 @@
 
-        ##settings string in json format
-        default_settings = KratosMultiphysics.Parameters("""
-        {
-            "solver_type"                  : "shallow_water_base_solver",
-            "model_import_settings"        : {
-                "input_type"          : "mdpa",
-                "input_filename"      : "unknown_name"
-            },
-            "echo_level"                   : 0,
-            "convergence_echo_level"       : 1,
-            "solver_echo_level"            : 0,
-            "buffer_size"                  : 2,
-            "dynamic_tau"                  : 0.005,
-            "dry_height"                   : 0.01,
-            "relative_tolerance"           : 1e-6,
-            "absolute_tolerance"           : 1e-9,
-            "maximum_iterations"           : 20,
-            "compute_reactions"            : false,
-            "reform_dofs_at_each_step"     : false,
-            "calculate_norm_dx"            : true,
-            "move_mesh_flag"               : false,
-            "volume_model_part_name"       : "volume_model_part",
-            "skin_parts"                   : [""],
-            "no_skin_parts"                : [""],
-            "linear_solver_settings"       : {
-                    "solver_type"     : "SkylineLUFactorizationSolver"
-            },
-            "time_stepping"                : {
-                "automatic_time_step" : false,
-                "time_step"           : 0.01
-            },
-            "pfem2_settings"               : {
-                "convection_scalar_variable"    : "HEIGHT",
-                "convection_vector_variable"    : "VELOCITY",
-                "maximum_number_of_particles"   : 16
-            }
-        }""")
-        default_settings["pfem2_settings"]["maximum_number_of_particles"].SetInt(8*self.domain_size)
+class ShallowWaterBaseSolver(PythonSolver):
+    def __init__(self, model, custom_settings):  # Constructor of the class 
+        settings = self._ValidateSettings(custom_settings)
+        
+        super(ShallowWaterBaseSolver, self).__init__(model, settings)
+
+        # There is only a single rank in OpenMP, we always print
+        self._is_printing_rank = True
+
+        ## Set the element and condition names for the replace settings
+        ## These should be defined in derived classes
+        self.element_name = None
+        self.condition_name = None
+        self.min_buffer_size = 2
+
+        # Either retrieve the model part from the model or create a new one
+        model_part_name = self.settings["model_part_name"].GetString()
+
+        if model_part_name == "":
+            raise Exception('Please specify a model_part name!')
+
+        if self.model.HasModelPart(model_part_name):
+            self.main_model_part = self.model.GetModelPart(model_part_name)
+        else:
+            self.main_model_part = KratosMultiphysics.ModelPart(model_part_name)
+            self.model.AddModelPart(self.main_model_part)
```

Comment:
Note:
If you do this here, i.e. before reading the ModelPart, then the SubmodelParts will only be accessible by "MainModelPartName.SubModelpartName" etc. whenever they are requested in the Model
This happens e.g. when the Processes are constructed.

This is the way to go in the future, but so far the names of the SubModelparts are usually not having also the MainModelPartName

I am not saying you shouldn't do this, but be aware it might break examples

=> In Fluid & Structure the ModelPart is added to the Model AFTER reading it, i.e. in `PrepareModelPart` to have a temp solution to not break existing examples

Reply:
Thank you for the two comments. I was forgetting to update an example.
I prefer to access the SubmodelParts by the full name ("main.sub"), like the way to go in the future.